src/Stream/BidirectionalReader.h

```diff
@@ -7,6 +7,11 @@ namespace OP2Utility::Stream
 {
 	class BidirectionalReader : public ForwardReader {
 	public:
+		inline void Peek(void* buffer, std::size_t size) {
+			ReadImplementation(buffer, size);
+			SeekBackward(size);
+		}
```

Comment:
Is there a reason this was explicitly declared [`inline`](https://en.cppreference.com/w/cpp/language/inline)? I believe that's already implied in this context.
> A function defined entirely inside a class/struct/union definition, whether it's a member function or a non-member friend function, is implicitly an inline function.

----

This method is rather low level, in that it takes an explicit buffer size, much like the internal read method. This makes it potentially error prone to use. It is unlike the externally used read template methods, which take a `T&` and use `sizeof` internally to determine the buffer size.

I'm not sure we really need a low level version of this method.

Actually the corresponding low level read method is `protected`, rather than `public`, so we probably want to avoid introducing a `public` low level peek method.

Reply:
Err, I'm pretty sure the corresponding function is public. See line 29 of Reader.h unless I'm reading it wrong.

```C++
inline void Read(void* buffer, std::size_t size) {
	ReadImplementation(buffer, size);
}
```

Do you want to make the original function protected?

I repeated the style when using the inline keyword from Reader.h. Guess I don't really have a preference. Since you are bringing up I'll assume you would prefer we not use the inline keyword in this sort of context.

Reply:
Ahh, right. I was looking at line 17.

I remember now, we had an internal function that was named differently to avoid name hiding in derived classes when just that one implementation method was overridden.

I suppose it does make sense to offer the public low level `Peek` function then.

I'm also seeing what you mean about multiple template methods. For some reason I was thinking it was just one template method. I forgot there were various restrictions and optimizations, which we handled with multiple overloads.

I wonder if there is a way to reduce overloads by using `constexpr if`. Or if there is a way to factor out those templates so they could be applied to both `Read` and `Peek`. That's starting to sound like future work though.

----

Hmm, that's true, we do use `inline` in a few places, and use it inconsistently. I probably didn't know what I was doing back then. We should probably look at removing those.

Reply:
Now that I think about it, I would liberally use inline in the past and probably started the precedence. I'll add an issue to remove inline keyword use. I think leeor had commented a while back that it was not needed for me to use inline so much as well...

Just added a trivial version of Peek. 

I think for now I'm happy with have 2 overloads for Peek. I was trying to think of Peek use cases. Typically, I am using it to check for something like the file type or maybe an integer representing the size of a data header tag. I don't know that it is super useful to Peek at full containers or the other more complex Read overloads. You would probably just be reading data in those situations.

Reply:
That's a good point about `Peek`. Usually it's to check for fixed sized sentinel values.